include/ginkgo/core/solver/workspace.hpp

```diff
@@ -0,0 +1,196 @@
+    }
+
+    template <typename ValueType>
+    Array<ValueType>& get()
+    {
+        GKO_ASSERT(this->template contains<ValueType>());
+        return dynamic_cast<concrete_container<ValueType>*>(data_.get())->array;
+    }
+
+    template <typename ValueType>
+    const Array<ValueType>& get() const
+    {
+        GKO_ASSERT(this->template contains<ValueType>());
+        return dynamic_cast<const concrete_container<ValueType>*>(data_.get())
+            ->array;
+    }
+
+private:
+    struct generic_container {
+        virtual ~generic_container() = default;
+    };
+
+    template <typename ValueType>
+    struct concrete_container : generic_container {
+        template <typename... Args>
+        concrete_container(Args&&... args) : array{std::forward<Args>(args)...}
+        {}
+
+        Array<ValueType> array;
+    };
+
+    std::unique_ptr<generic_container> data_;
+};
+
+
+class workspace {
+public:
+    workspace(std::shared_ptr<const Executor> exec) : exec_{std::move(exec)} {}
+
+    workspace(const workspace& other) : workspace{other.get_executor()} {}
+
+    workspace(workspace&& other) : workspace{other.get_executor()}
+    {
+        other.clear();
+    }
+
+    workspace& operator=(const workspace& other) { return *this; }
+
+    workspace& operator=(workspace&& other)
+    {
+        other.clear();
+        return *this;
+    }
+
+    template <typename VectorType, typename CreateOperation>
+    VectorType* create_or_get_vector(int vector_id, CreateOperation op,
+                                     const std::type_info& expected_type,
+                                     dim<2> size, size_type stride)
+    {
+        if (vector_id >= vectors_.size()) {
+            vectors_.resize(vector_id + 1);
+        }
+        // does the existing object have the wrong type?
+        // vector types may vary e.g. if users derive from Dense
+        auto stored_vec = vectors_[vector_id].get();
+        VectorType* vec{};
+        if (!stored_vec || typeid(*stored_vec) != expected_type) {
+            auto new_vec = op();
+            vec = new_vec.get();
+            vectors_[vector_id] = std::move(new_vec);
+            return vec;
+        }
+        // does the existing object have the wrong dimensions?
+        vec = dynamic_cast<VectorType*>(vectors_[vector_id].get());
+        GKO_ASSERT(vec);
+        if (vec->get_size() != size || vec->get_stride() != stride) {
+            auto new_vec = op();
+            vec = new_vec.get();
+            vectors_[vector_id] = std::move(new_vec);
+        }
+        return vec;
+    }
+
+    const LinOp* get_vector(int vector_id) const
+    {
+        GKO_ASSERT(vector_id < vectors_.size());
+        return vectors_[vector_id].get();
+    }
+
+    template <typename ValueType>
+    Array<ValueType>& create_or_get_array(int array_id, size_type size)
+    {
+        if (array_id >= arrays_.size()) {
+            arrays_.resize(array_id + 1);
+        }
```

Comment:
resize explicitly set the capacity, so it keeps change the size of vector from 1 to total number.
Use emplace_back to let vector-self decide memory?

Reply:
the array entries are tiny, and if I use emplace_back, I lose control of the vector IDs, especially if I accidentally change their order.